app/models/journable/with_historic_attributes.rb

```diff
@@ -0,0 +1,212 @@
+#
+#   # Simplified interface for two timestamps
+#   query.timestamps  # => [<Timestamp 2022-01-01T00:00:00Z>, <Timestamp PT0S>]
+#   work_package = Journable::WithHistoricAttributes.wrap(work_package, query:)
+#   work_package.baseline_timestamp  # => [<Timestamp 2022-01-01T00:00:00Z>]
+#   work_package.current_timestamp  # => [<Timestamp PT0S>]
+#   work_package.matches_query_filters_at_baseline_timestamp?
+#   work_package.matches_query_filters_at_current_timestamp?
+#   work_package.baseline_attributes.subject # => "Subject at 2022-01-01 (baseline time)"
+#   work_package.subject  # => "Subject at PT0S (current time)"
+#
+class Journable::WithHistoricAttributes < SimpleDelegator
+  attr_accessor :timestamps,
+                :query,
+                :include_only_changed_attributes,
+                :loader
+
+  def initialize(journable,
+                 timestamps: nil,
+                 query: nil,
+                 include_only_changed_attributes: false,
+                 loader: Loader.new(journable))
+    super(journable)
+
+    if query and not journable.is_a? WorkPackage
+      raise Journable::NotImplementedError, "Journable::WithHistoricAttributes with query " \
+                                            "is only implemented for WorkPackages at the moment " \
+                                            "because Query objects currently only support work packages."
+    end
+
+    self.query = query
+    self.timestamps = timestamps || query.try(:timestamps) || []
+    self.include_only_changed_attributes = include_only_changed_attributes
+
+    self.loader = loader
+  end
+  private_class_method :new
+
+  class << self
+    def wrap(journable_or_journables,
+             timestamps: query.try(:timestamps) || [],
+             query: nil,
+             include_only_changed_attributes: false)
+      wrapped = wrap_each_journable(Array(journable_or_journables), timestamps:, query:, include_only_changed_attributes:)
+
+      case journable_or_journables
+      when Array, ActiveRecord::Relation
+        wrapped
+      else
+        wrapped.first
+      end
+    end
+
+    private
+
+    def wrap_each_journable(journables, timestamps:, query:, include_only_changed_attributes:)
+      loader = Loader.new(journables)
+      journables = loader.at_timestamp(timestamps.last).values if timestamps.last.try(:historic?)
+
+      journables.map { |j| new(j, timestamps:, query:, include_only_changed_attributes:, loader:) }
+    end
+  end
+
+  def attributes_by_timestamp
+    @attributes_by_timestamp ||= Hash.new do |h, t|
+      attributes = if include_only_changed_attributes
+                     changes_at_timestamp(t)&.transform_values(&:last)
+                   else
+                     historic_attributes_at(t)
+                   end
+
+      h[t] = attributes ? Hashie::Mash.new(attributes) : nil
+    end
+  end
+
+  def changed_at_timestamp(timestamp)
+    changes_at_timestamp(timestamp)&.keys || []
+  end
```

Comment:
I chose the name because it is similar to `ActiveModel::Dirty#changed` with the addition of `at_timestamp`. I therefore wouldn't want to change the name as I think it quite good for somebody being familiar with the `Dirty` module. I rather added documentation in 8495bff2d7.